README
WWW::Myspace - Access MySpace.com from Perl scripts

WARNING

March 2007: Using WWW::Myspace for commenting, messaging, or adding
friends will probably get your Myspace account deleted or disabled.


WWW::Myspace.pm provides methods to access myspace.com accounts and functions
automatically. It provides a simple interface for scripts to log in,
access lists of friends, scan user's profiles, retreive profile
data, and post comments.

Example:

    use WWW::Myspace;
    my $myspace = WWW::Myspace->new ($account, $password);
        OR
    my $myspace = new WWW::Myspace; # Prompts for email and password

    my ( @friends ) = $myspace->get_friends();


PREREQUESITES

    (Makefile.PL will always have the the current list. This was copied from there Jun 6, 2007)

        'Test::More' => 0,
        'Spiffy' => 0.24,
        'YAML' => 0.39,
        'File::Spec::Functions' => 0, # Note: Forces File::Spec 0.7 or later.
        'Contextual::Return'    => 0, # For send_friend_request method
        'Locale::SubCountry'    => 0, # FriendAdder.pm, cool_new_people
        'WWW::Mechanize'        => 1.20, # Myspace.pm
        'Time::Local'			=> 0, # Myspace.pm - last_login method
        'Crypt::SSLeay'			=> 0.53, # WWW::Mechanize, for SSL access to myspace.com
        'Config::General'       => 0, # MyBase.pm
        'Params::Validate'      => 0, # MyBase.pm


Each of them has a list of dependencies also.

If you don't have XML::Parser installed, you may need to manually install expat:
http://sourceforge.net/projects/expat/
Just try installing and if the install fails look through the messages
and see if it says you need expat.

On MacOS X you'll need to install the developer tools from your
Tiger install DVD or download them from http://developer.apple.com
or nothing will install and you'll get errors about missing CORE files).
(I haven't tested this on pre-Tiger MacOS, but it should work with
those developer tools also).

INSTALLATION

Installing using the cpan command-line tool is the easiest
as it will install the dependencies for you.

cpan install WWW::Myspace

or

sudo cpan install WWW::Myspace

Good general advice for installing any CPAN module:
If a module fails to install:
- Starting at the top, scan the copious output cpan will create
- When you see a module has failed to install:
  - cd into the module's directory ( hint: cd ~/.cpan/build ; ls )
  - Read the README for the module and install it manually. You'll
    see the specific errors it's having there.
  - Once it's installed, run your cpan install command again.

If you don't configure cpan to use "sudo make install"
(and you should, for which you'll need cpan 1.83), you'll probably
want to run the cpan command as root or run "sudo cpan". man cpan
for more info.


To install this module manually, run the following commands:

    perl Makefile.PL
    make
    make test
    sudo make install  # Or su root ; make install


The distribution also installs the following convenience modules:
WWW::Myspace::Comment
WWW::Myspace::FriendChanges
WWW::Myspace::Message

Also be sure to check out:
WWW::Myspace::FriendAdder
WWW::Myspace::Data
which are available as separate modules (these used to be included
but were separated out on 6/29/2007 to make the base module lighter).

The distribution also installs the following handy scripts:
approve_friends
comment_myspace
message_group

(add_friends was moved to WWW::Myspace::FriendAdder)

These are scripts I use, and decided to include in the
distribution in case they're handy, since they're the most common
uses for the module, and they're a bit more robust and friendly
than just using the methods in the module themselves.


TEST SUITE

WWW::Myspace accesses a web site, which by nature may change.
Such change may break functions in this module.
The test suite (run during cpan installation or when you type
"make test") runs an extensive battery of tests on all methods
against the myspace site. This has a few effects:
1) It makes sure the module works on your system.
2) It makes sure that the module works with the current version
   of the myspace site.
3) If the site changes, or has problems, tests may fail even though
   overall the module will work.
My approach to programming is that the module should be robust
enough to handle the changes in the myspace site, so if condition
3 occurs, the module needs beefing up.

Note that you can re-run the test suite at any time. Keep a copy
of the distribution (hint: cd ~/.cpan/build/WWW-Myspace* ) and
run:
    perl Makefile.PL
    make test

This is handy if you find that something in one of your scripts
stops working and you're wondering if something changed that broke
WWW::Myspace. If the tests pass, it's probably your script. If not,
there'll probably be an update coming soon (make sure you have
the latest version, of course - just re-run the cpan command).

IMPORTANT NOTE ABOUT TESTS: As of version 0.40, you can run the full
test suite against your own accounts.  To do so, place a config
file (copy/paste this and change the values) in
~/.www-myspace/test_config.yaml:

---
acct1:
  friend_id: 12345
  username: myspace_login_email_address
  password: myspace_login_password
  user_name: Username (as it appears in "Hello, Username" when you log in.)
acct2:
  friend_id: 
  username: 
  password: 
  user_name: 
test_group: 100009984


See t/config.yaml for an example file (but lacking "username" and "password").

This establishes test data for 2 accounts, "acct1", and "acct2".  You provide
the login info, friendID, and display name for each account. The test
scripts then log into each account and test the module's methods using them.
They'll post comments back and forth, send a message or two, and delete and
add each other as friends.  It should be safe to run on an active account,
but if you feel nervous, set up a couple dummy accounts.

NOTE THAT acct2 MUST BE A PERSONAL (not a band) ACCOUNT, or the is_band
test will fail, causing your tests to fail.


HOW TO SUBMIT A PATCH

To submit a patch for a new feature or a bug fix, please observe the following.
Doing so will allow me to implement your patch quickly.  Not doing so may
delay its implementation or prevent me from implementing your patch at all.

 - Download the newest development version from SVN.
   The command to use is here:
   http://sourceforge.net/svn/?group_id=163042
   (Or see http://sourceforge.net/projects/www-myspace)
 - Apply your changes to that version.
 - Create a unified or context diff of the changed file(s):
   diff -u original_file.pm your_file.pm
 - Email the output to bug-www-myspace at rt.cpan.org, or go to the
   CPAN RT web site (see below) and submit it there.

I will apply your patch and run the tests on it.


SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with
the perldoc command.

  perldoc WWW::Myspace

You can also look for information at:

    Search CPAN
        http://search.cpan.org/dist/WWW-Myspace

    CPAN Request Tracker:
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Myspace

    AnnoCPAN, annotated CPAN documentation:
        http://annocpan.org/dist/WWW-Myspace

    CPAN Ratings:
        http://cpanratings.perl.org/d/WWW-Myspace

COPYRIGHT AND LICENCE

This module copyright (C) 2005-2006 Grant Grueninger except
FriendAdder.pm, (c) 2006 Olaf Alders.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.


Changes
Revision history for WWW::Myspace

0.91    2008-11-20
        - Fixed approve_friend_requests which was no longer working as a result
          of Myspace changes (RT#41074, patch from David Moreno Garza)
        - Example script samples/approve_friends also fixed by the above change

0.90    2008-11-15
        - Fixed a new problem in get_inbox and read_message, which were no
          longer returning the message sender's name or ID due to Myspace
          changes probably made on 2008-11-15

0.89    2008-11-14
        - This release replaces a dud re-release of 0.88 made earlier today
          which had some of the following enhancements but the version
          number had not been increased
        - Fixed a problem with get_inbox whereby no messages were returned
          unless the inbox held more than one page of messages (more than 10
          messages)
        - Fixed post_comment, broken by recent Myspace changes
        - Altered CAPTCHA handling in post_comment such that the configured
          handler (if any) is now called repeatedly if CAPTCHA responses are
          rejected

0.88    2008-10-01
        - Fixed is_band to correctly detect music profiles, which broke due to
          Myspace changes around 2008-09-30;  this was also causing
          get_basic_info to fail to work with music profiles
        - Fixed get_basic_info to return correct timestamps in 'lastlogin'
          field, both for personal and music profiles
        - New tests in t/17-get_basic_info.t check that that get_basic_info
          returns a sensible 'lastlogin' timestamp for personal and music
          profiles, and a sensible 'profileviews' for a band profile
        - New tests in t/22-search_music.t check that multiple results pages
          are read, and that spaces in search terms are properly escaped

0.87    2008-09-12
        - New method captcha_handler allows a user-defined function to provide
          captcha responses directly
        - For backward-compatibility, setting captcha_killer_api_key will
          automatically enable the built-in Captcha Killer handler
        - Fixed send_friend_request to submit CAPTCHA responses correctly
        - Fixed send_message to submit CAPTCHA responses correctly
        - Fixed post_comment to submit CAPTCHA responses correctly
        - Fixed the login mechanism to return an error upon a failed login
          attempt.  An invalid username/password should also now return the
          appropriate error message.
        - Added to the documentation of the 'new' method (the constructor) a way
          to detect an invalid username/password error within a script
        - Restored the t/11-badlogin.t test

0.86    2008-09-10
        - Fixed get_friends;  in most cases the method would only have worked
          the first time it was executed (fixed RT#38716)
        - Created new tests t/21-get-friends.t for get_friends method
        - Fixed submit_form flaw;  fields_ref entries were not being submitted
          where a form input of the same name existed but with the disabled
          attribute set
        - Fixed search_music;  searching by keywords was not working due to the
          above flaw in submit_form, and the method was looking for search
          results beyond the last page.  Also now croaks if given a parameter
          which is not a hash ref.
        - Created new tests t/22-search-music.t for search_music method
        - Added to send_message documentation a note about line endings and
          accounts which disable HTML in incoming messages

0.85    2008-08-23
        - Fixed get_inbox;  method was always returning zero messages due to
          changes around 2008-08-20 (RT#38639)

0.84    2008-08-16
        - Fixed last_login to ignore a supplied 'undef' parameter (RT#38457)
        - Fixed problems in test script t/01-login.t which caused that script to
          abort without running all tests;  there was inadequate error checking
          when opening a data file, and the file t/login-forms/latest.html was
          missing from the MANIFEST in the 0.83 release

0.83    2008-08-15
        - Fixed send_message, broken by Myspace changes 2008-08-15 (RT#38494)
        - Fixed send_friend_request, broken by Myspace changes (RT#38385)
          of Myspace changes
        - Fixed friends_from_profile;  function was returning friends from
          the logged-in account, instead of the profile specified (RT#37057)
        - Removed example documenting the use of the "exclude" option
          from friends_from_profile method.  This option relied on the
          "exclude" option in get_friends, which was removed during the
          re-write of get_friends_on_page in version 0.62.
        - Fixed get_basic_info
        - Fixed friend_user_name
        - Fixed last_login
        - Fixed last_login_ymd
        - read_message now cleans the 'gauntlet_tokenizer_reserved' attribute
          from HTML tags in received messages
        - Disabled cool_new_people function and tests because the feature has
          been replaced by Myspace with a similar feature on the homepage;  this
          may be supported in the future
        - It should be no longer necessary to change an account's language/
          location to English/United States;  this is now handled automatically
          without changing your account settings directly (currently by 
          modifying the MSCulture cookie)
        - Much functionality previously worked with language/location set to
          English/United Kingdom, until changes on 2008-07-18 which broke
          messaging functions.  That is fixed in this release by way of the
          change mentioned above
        - Send Accept-Language HTTP header in anticipation of Myspace one day
          respecting it when guessing the visitor's language preference
        - Added a file 'mechanize.patch' which can be used to stop warnings

Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
#    (MM->can('signature_target') ? (SIGN => 1) : ()),
    NAME                => 'WWW::Myspace',
    AUTHOR              => 'Grant Grueninger <grantg@cpan.org>',
    VERSION_FROM        => 'lib/WWW/Myspace.pm',
    ABSTRACT_FROM       => 'lib/WWW/Myspace.pm',
    PL_FILES            => {},
    'EXE_FILES'         => [qw( scripts/approve_friends
                                scripts/comment_myspace
                                scripts/message_group
                              )],
    PREREQ_PM => {
        'Test::More' => 0,
#        'HTTP::Request::Form' => 0,
        'Spiffy' => 0.24,
        'YAML' => 0.39,
#        'IO::All' => 0.33, # Needed for the scripts
        'File::Spec::Functions' => 0, # Note: Forces File::Spec 0.7 or later.
        'Contextual::Return'    => 0, # For send_friend_request method
        'Locale::SubCountry'    => 1.38, # FriendAdder.pm, cool_new_people
        'WWW::Mechanize'        => 1.20, # Myspace.pm
#        'Time::Local'			=> 0, # Myspace.pm - last_login method
        'Crypt::SSLeay'			=> 0.53, # WWW::Mechanize, for SSL access to myspace.com
        'Config::General'       => 0, # MyBase.pm
        'Params::Validate'      => 0, # MyBase.pm
        'Time::ParseDate'       => 100.010301, # Myspace.pm - last_login method.
        'URI::Escape'           => 0 # Myspace.pm - find_friend method
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'WWW-Myspace-*' },
);
